### PR TITLE
Modernise tooling and fix CI failing

### PR DESCRIPTION
**Pull request by DMRobertson**
Commitwise reviewable.

I think the only necessary changes were

- Fix a unit test failure
- Dump flake8 for ruff

In particular, moving to pyproject.toml was my first attempt/guess to fix the CI errors, but I don't think it was strictly necessary.

### CODE REVIEW

**Comment by DMRobertson** [comment]
This is an exc

> Failing flake8 looks like this: [python/importlib_metadata#406](https://github.com/python/importlib_metadata/issues/406)
> 
> :(

Good spot, thank you.

This is an excellent excuse for us to move to ruff instead of flake8.

File: scripts-dev/lint.sh [hunk 1]
```diff
@@ -98,5 +98,5 @@ set -x
 
 isort "${files[@]}"
 python3 -m black "${files[@]}"
-flake8 "${files[@]}"
+ruff "${files[@]}"
```
Comment (clokep) [inline scripts-dev/lint.sh:101]:
```suggestion
ruff --quiet --fix "${files[@]}"
```

File: .github/workflows/pipeline.yml [hunk 2]
```diff
@@ -30,24 +30,31 @@ jobs:
...
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
         with:
           python-version: "3.7"
-      - name: Install old dependencies
-        run: pip install -r <(./scripts-dev/old_deps.py)
+      - name: Patch pyproject.toml to require oldest dependencies
+        run: >
+          # Ugly. Could use something like https://pyproject-parser.readthedocs.io/en/latest/cli.html#info in the future.
+          sed -i -e '' pyproject.toml
```
Comment (clokep) [inline .github/workflows/pipeline.yml:52]:
```suggestion
          sed -i -e 's/>=/==/g' pyproject.toml
```

Comment (DMRobertson) [reply .github/workflows/pipeline.yml:52]: Whoops! Thank you!

**Comment by dklimpel** [comment]
> Fix an unintended rename in

Sorry.

**Comment by DMRobertson** [comment]
> > Fix an unintended rename in #333
> 
> Sorry.

No worries. It's easily done, and was our fault for having  broken CI.

**Review by clokep — APPROVED** [review]

**Pull request merged**